lib/Target/AVR/AVRInstrInfo.cpp

```diff
@@ -45,6 +45,30 @@ extern bool RA_InSpillerCode;
 AVRInstrInfo::AVRInstrInfo()
     : AVRGenInstrInfo(AVR::ADJCALLSTACKDOWN, AVR::ADJCALLSTACKUP), RI() {}
 
+// FIXME: BEFORE MERGING: Is there a cleaner tablegen method of doing this?
```

Comment:
Thinking harder on this, I don't quite understand what the generated code is doing. This LLVM IR

```llvm
define i1 @unsigned_multiplication_did_overflow(i8, i8) unnamed_addr {
entry-block:
  %2 = tail call { i8, i1 } @llvm.umul.with.overflow.i8(i8 %0, i8 %1)
  %3 = extractvalue { i8, i1 } %2, 1
  ret i1 %3
}
```

Generates this AVR assembly:

```asm
unsigned_multiplication_did_overflow:
                         ;; 8-bit inputs seem to come as 16-bit values
        mul     r24, r22 ;; Multiply the low bytes; result is in r1:r0
        mov     r24, r1
        eor     r1, r1   ;; Inefficient - no need to reset zero register
        cpi     r24, 0   ;; Check if the high byte was zero
        brne    LBB0_2
        ldi     r18, 0   ;; No overflow, return 0 as 16-bit value
        ldi     r19, 0
        mov     r24, r19 ;; ??? Why are we doing this ???
        ret
LBB0_2:                  ;; It was not zero; we had overflow
        ldi     r18, 1
        ldi     r19, 0   ;; Return 1 as a 16-bit value
        mov     r24, r19 ;; ??? Why are we doing this ????

```

This seems to make sense up until the end — why are those rogue `mov`s at the end? Those are the ones that generate the move from a 16-bit register to an 8-bit one. That seems **super** suspicious.